bap-vibes/src/patch_c.ml

```diff
@@ -0,0 +1,1501 @@
+        pre;
+        post;
+        ASSIGN (tmp, e2);
+      ] in
+    let+ e = f op e1 (VARIABLE tmp) t in
+    eff, Some e, NOP
+  | Second (pre1, e1, post1), Second (pre2, e2, post2) ->
+    let eff = sequence [
+        pre1;
+        post1;
+        pre2;
+        post2;
+      ] in
+    let+ e = f op e1 e2 t in
+    eff, Some e, NOP
+
+(* Translate binary operators. *)
+and translate_binary_operator
+    (b : Cabs.binary_operator)
+    (lhs : Cabs.expression)
+    (rhs : Cabs.expression) : (stmt * exp option * stmt) transl =
+  let exp = translate_expression_strict "translate_binary_operator" in
+  let default ?(no_ptr = false) op =
+    let* spre1, e1, spost1 = exp lhs in
+    let t1 = typeof e1 in
+    let* spre2, e2, spost2 = exp rhs in
+    let t2 = typeof e2 in
+    match typ_unify t1 t2 with
+    | None -> typ_unify_error Cabs.(BINARY (b, lhs, rhs)) t1 t2
+    | Some (PTR _) when no_ptr ->
+      typ_unify_error Cabs.(BINARY (b, lhs, rhs)) t1 t2
+    | Some t ->
+      let e1 = with_type e1 t in
+      let e2 = with_type e2 t in
+      let f op e1 e2 t = Transl.return @@ BINARY (op, e1, e2, t) in
+      binary_tmp_or_simple op spre1 e1 spost1 spre2 e2 spost2 t ~f
+        ~no_post:true in
+  match b with
+  | Cabs.ADD -> translate_arith ADD b lhs rhs
+  | Cabs.SUB -> translate_arith SUB b lhs rhs
+  | Cabs.MUL -> translate_arith MUL b lhs rhs ~no_ptr:true
+  | Cabs.DIV -> translate_arith DIV b lhs rhs ~no_ptr:true
+  | Cabs.MOD -> translate_arith MOD b lhs rhs ~no_ptr:true
+  | Cabs.AND -> translate_short_circuit_and lhs rhs
+  | Cabs.OR -> translate_short_circuit_or lhs rhs
+  | Cabs.BAND -> default LAND ~no_ptr:true
+  | Cabs.BOR -> default LOR ~no_ptr:true
+  | Cabs.XOR -> default XOR ~no_ptr:true
+  | Cabs.SHL -> default SHL ~no_ptr:true
+  | Cabs.SHR -> default SHR ~no_ptr:true
+  | Cabs.EQ -> default EQ
+  | Cabs.NE -> default NE
+  | Cabs.LT -> default LT
+  | Cabs.GT -> default GT
+  | Cabs.LE -> default LE
+  | Cabs.GE -> default GE
+  | Cabs.ASSIGN -> translate_assign lhs rhs
+  | Cabs.ADD_ASSIGN -> translate_compound ADD b lhs rhs
+  | Cabs.SUB_ASSIGN -> translate_compound SUB b lhs rhs
+  | Cabs.MUL_ASSIGN -> translate_compound MUL b lhs rhs ~no_ptr:true
+  | Cabs.DIV_ASSIGN -> translate_compound DIV b lhs rhs ~no_ptr:true
+  | Cabs.MOD_ASSIGN -> translate_compound MOD b lhs rhs ~no_ptr:true
+  | Cabs.BAND_ASSIGN -> translate_compound LAND b lhs rhs ~no_ptr:true
+  | Cabs.BOR_ASSIGN -> translate_compound LOR b lhs rhs ~no_ptr:true
+  | Cabs.XOR_ASSIGN -> translate_compound XOR b lhs rhs ~no_ptr:true
+  | Cabs.SHL_ASSIGN -> translate_compound SHL b lhs rhs ~no_ptr:true
+  | Cabs.SHR_ASSIGN -> translate_compound SHR b lhs rhs ~no_ptr:true
+
+(* Captures the short-circuiting semantics of the && operator. *)
+and translate_short_circuit_and
+    (lhs : Cabs.expression)
+    (rhs : Cabs.expression) : (stmt * exp option * stmt) transl =
+  let exp = translate_expression_strict "translate_short_circuit_and" in
```

Comment:
since 4.12 we can use `__FUNCTION__` to get the enclosing function name.

Reply:
I'm using 4.11.2 right now, so we would need to bump the required OCaml version for VIBES.